### PR TITLE
I302 responsive tables

### PR DESCRIPTION
**Pull request by GaryRidgway**
This PR is to work on responsive tables. This is a big thing so i would really like some review if possible If you want to test bot the functionality of tables and their usage in `uids_base`, i suggest starting with the steps in the `To test in uids_base` section.

Browser testing would probably be nice, as I am not sure how many different browsers this has been tested in.

To test in uids:
1. get uids set up and check out this branch
2. run `gulp watch` in the base uids folder
3. check the tables component and make sure that it is working with horizontal and vertical sticky headers. This can be done by making the viewport small and seeing if the headers travel with the viewport.
4. change the table data around so there is only headers in the horizontal, or vertical, and make sure it still works.
5. check that all `<th>` tags are receiving their respective scope tags and that an accessible wrapper is being put around the table.
6. check that you can disable all sticky functionality by adding the `table-static` class to the `table` element
7. check that the sticky top-left corner only works when you have both column AND row headers

To test in uids base:
1. pull down a site using uids base and checkout the branch `i2137-responsive-tables`, open the `package.json` file in `/docroot/themes/uids_base` and under the `dependencies` header, change the `@uiowa/uids` entry to `uiowa/uids#c0c9976`, which should be the most recent commit hash for the branch this is being worked on.
    * (I sometimes had to change the `@uiowa/uids` entry first to `uiowa/uids#i302-responsive-tables` in order for `blt frontend` to grab the `tables.js` file. Once it was in the directory, `uiowa/uids#983f78e` worked. I'm guessing creating a blank `tables.js` file prior to `blt frontend` would also work. - Cory)
2. run a `blt frontend` for it to copy over the branches js and scss and compile them.
3. `blt dsa` a layout builder or paragraphs site and create a new page.
4. make a bunch of sections with differing numbers of columns and place tables in them. you should preferably test tables with both the `.is-striped` class and without it. (ex:  `<table class="table is-striped">` and `<table class="table">`)
5. when viewing the page, all of the tables should function as stated in steps 3-7 from the `To test in uids:` section. As well, they should resize to the column widths when you expand and contract the viewport width.

<details>
<summary>This is some dummy table markup I have been using to make my life easier. Insert it in source editing mode in the WYSIWYG for easy table action!</summary>
<p>

```HTML
<table class="is-striped table">
	<caption>Hello!</caption>
	<thead>
		<tr>
			<th scope="row">Corner</th>
			<th scope="col">Column One</th>
			<th scope="col">Column Two</th>
			<th scope="col">Column Three</th>
			<th scope="col">Column Four</th>
			<th scope="col">Column Five</th>
			<th scope="col">Column Six</th>
			<th scope="col">Column Seven</th>
			<th scope="col">Column Eight</th>
			<th scope="col">Column Nine</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<th scope="row">Row One</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Two</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Three</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Four</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Five</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Six</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Seven</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Eight</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Nine</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
		<tr>
			<th scope="row">Row Ten</th>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
			<td>187 Fifth Age</td>
		</tr>
	</tbody>
</table>

### CODE REVIEW

**Review by dnkemper — DISMISSED** [review]

**Comment by joewhitsitt** [comment]
I took a glance at: https://uids.brand.uiowa.edu/branches/i302-responsive-tables/components/preview/tables.html

I think this works pretty well for a well crafted table. The header is delayed in catching up with user-scroll but I imagine there is a technical limitation there since it is separate.

I think this should be presented at the design meeting before it get merged in.

**Comment by bspeare** [comment]
Looks good!  A couple of quick thoughts. 

- Maybe make the divider lines on the top example for https://uids.brand.uiowa.edu/branches/i302-responsive-tables/components/preview/tables.html a lighter gray?
- Should we not force uppercase on the vertical left column?
- Maybe soften the box shadow a step further?

**Comment by bspeare** [comment]
One more thing, maybe only add the shadow to the top table header only on mobile scroll?  I know we briefly discussed that option.   For desktop view, I'd prefer to see it without the shadow.

**Comment by mark-bennett-uiowa** [comment]
@GaryRidgway Scrolled through the code and things look great from that standpoint. I do agree with @bspeare above in that I feel we shouldnt be forcing uppercase on (any) headings. It looks like src/components/tables/tables.scss lines 52 and 151 still has this. Im not a big fan of them personally and when I looked up some details on any standards, it looks like this has been some legibility research on it: [https://ncbi.nlm.nih.gov/pmc/articles/PMC2016788](https://ncbi.nlm.nih.gov/pmc/articles/PMC2016788). I know I cherry-picked this article, so if there is any desire from the rest of the group to keep the uppercase, I can follow.

Other than my small concern above, your work on this has been AMAZING. I can test this in greater detail if I get a chance, or if someone doesnt beat me to it.

**Comment by cory-skeers** [comment]
Having trouble narrowing down what's causing it, but occasionally running into this issue where a duplicate horizontal header is getting rendered, and it breaks the containers.

Screen Shot 2020-09-14 at 6 36 34 PM

On layouts with a small column (4col layout here, with 25% column widths), it's possible to end up with a row header that takes up most or all of the column space while remaining sticky, hiding the rest of the table.

Example shown is in an 1100px viewport, 4col layout with a long single-word row header. (Definitely an edge case, and the straightforward workaround is to not try to include four tables with wide labels in a single section...)
image

**Comment by cory-skeers** [comment]
> * Having trouble narrowing down what's causing it, but occasionally running into this issue where a duplicate horizontal header is getting rendered, and it breaks the containers.

Looks like this is coming from tables w/out a caption. Doesn't always lead to both headers being visible, but can depending on table size and placement. Guessing this accounts for what I was running into before.

image

**Comment by pyrello** [comment]
I am in the process of reviewing this.

**Comment by joewhitsitt** [comment]
Reviewed this from UIDS for now. Don't want to be the one to hold this up from moving forward. I threw what I thought was a fairly complex table (but that exists on one of our websites) at this and it is holding up. Additionally, the `.table-static` class works as I expect it to and that other custom classes I add aren't stripped!

The concerns I have are probably more with general uids styling (e.g. th borders only on right break when colspan differs) and not the functional aspects of responsive tables.

Screen Shot 2020-09-18 at 11 30 34 AM

I did notice that captions aren't hidden in uids and are hidden in uids_base. Without it in UIDS, the caption appears where it is in the head of the table and covered by the sticky header and there is a gap of borders. I wonder if we can find a place for the caption and NOT hide it. Possibly the bottom of the table like Bootstrap? Captions can be valuable to visual users as well... docroot/themes/custom/uids_base/scss/components/tables.scss https://getbootstrap.com/docs/4.5/content/tables/#captions

File: src/components/tables/tables.js
Comment (pyrello) [inline src/components/tables/tables.js:24]: We do not put Drupal-specific code in the UIDS implementation. This needs to be moved to a uids_base js file.

File: src/components/tables/tables.scss
Comment (pyrello) [inline src/components/tables/tables.scss:19]: This is a weird mixture of camelCase and kebab-case. Our class naming up to this point has been entirely kebab-case, so I think we should stick with that.

In this case, the line above would be re-written:
```
  &.is-sticky-left.left-sticky-minwidth {
```

And of course, all references to the classes in question would need to be updated.

File: src/components/tables/tables.scss
Comment (pyrello) [inline src/components/tables/tables.scss:59]: There is nothing about this class name that really helps me to understand what its relationship is to the rest of the table. I would like to see this re-written to take advantage of a BEM style class naming. So that it is more clear at a glance what is happening.

**Review by pyrello — CHANGES_REQUESTED** [review]
@GaryRidgway This is a great start. It needs some refinement for consistency and better readability. I'll have more feedback, but I wanted to let you know I think this needs some changes before it should be merged in.

**Comment by pyrello** [comment]
I converted this back to a draft state because it already had an approval which I could not find a way to remove. Since there is active refactoring going on, draft seems more appropriate.

**Comment by GaryRidgway** [comment]
I changed the structure of all classes and IDs in to BEM structure and added a better separation of javascript between uids and uids_base. a couple of key feature changes:

- the `table-static` class to disable responsive tables is now `table--static` to be more in line with BEM standards
- there is now a user definable hook called `hook_triggerTableRespond` that a user can define in a theme from any platform and hook in to the `triggerTableRespond function right at the beginning of it's call.

**Comment by bspeare** [comment]
@GaryRidgway is it possible to take advantage of the SASS nesting structure on this so that it reads more like:

```
.table {
  &__child-element {
  }
  &--modifier {
  }
}
```

File: src/components/tables/tables.scss
Comment (bspeare) [inline src/components/tables/tables.scss:63]: Is it possible to add a brief comment on what the `track-heading` is?  This is the only class that I wasn't sure about after reading through this.

**Comment by bspeare** [comment]
A section like https://github.com/uiowa/uids/blob/i302-responsive-tables/src/components/tables/tables.scss#L54-L64 could also take advantage of the SASS nesting.

**Comment by bspeare** [comment]
Also, maybe we should have the base `table` styles first and then the responsive stuff after in the order?

File: src/components/tables/tables.scss
Comment (GaryRidgway) [reply src/components/tables/tables.scss:63]: So actually i was debating on this one to make it larger and or more descriptive, however, i thought `.table__responsive-header__scroll-track__track-heading` felt really long? it doe add clarity but I was unsure of whether that was worth it or not. Should i change the class?

Comment (pyrello) [reply src/components/tables/tables.scss:63]: I think he means add to the comments.

Comment (pyrello) [reply src/components/tables/tables.scss:63]: @GaryRidgway I think this is actually too verbose. I think something like `.table__sticky-heading` is sufficient.

Comment (GaryRidgway) [reply src/components/tables/tables.scss:63]: Right, I was trying to start a conversation on the core of the issue. I was wondering if BEM was explaining it in the right way or not, and if a comment was sufficient, or if it was better to rearrange the BEM class syntax. I can surely just add the comment, but I wanted to ask this first to try and solve the core of the issue instead of just put a band-aid on it.

Comment (GaryRidgway) [reply src/components/tables/tables.scss:63]: @pyrello  sounds good!

Comment (pyrello) [reply src/components/tables/tables.scss:63]: @GaryRidgway Thanks for clarifying. I think that the BEM structure helps with readability, but comments are also necessary. Good discussion :)

**Comment by GaryRidgway** [comment]
Okiokey, this should be cleaned up! I changed a fair amount and combined and streamlined a bunch of styles.

**Comment by GaryRidgway** [comment]
testing in uids_base now, think something got goofed. trying to fix that

**Comment by bspeare** [comment]
Refactor looks good!  Thanks for the comment too!

**Review by bspeare — APPROVED** [review]
Looks good!

**Review by GaryRidgway — COMMENTED** [review]
this should all be resolved now with the latest update

**Pull request merged**
